contracts/RestrictedNORI.sol

```diff
@@ -0,0 +1,1047 @@
+  error InsufficientUnreleasedTokens(uint256 scheduleId);
+  error InsufficientClaimableBalance(address account, uint256 scheduleId);
+  error InvalidBpNoriSender(address account);
+  error InvalidScheduleStartTime(uint256 projectId);
+
+  /** The internal governing parameters and data for a schedule */
+  struct Schedule {
+    uint256 startTime;
+    uint256 endTime;
+    uint256 totalClaimedAmount;
+    bool exists;
+    uint256 totalQuantityRevoked;
+    uint256 releasedAmountFloor;
+    EnumerableSetUpgradeable.AddressSet tokenHolders;
+    mapping(address => uint256) claimedAmountsByAddress;
+    mapping(address => uint256) quantitiesRevokedByAddress;
+  }
+
+  /** View information for the current state of one schedule */
+  struct ScheduleSummary {
+    uint256 scheduleTokenId;
+    uint256 startTime;
+    uint256 endTime;
+    uint256 totalSupply;
+    uint256 totalClaimableAmount;
+    uint256 totalClaimedAmount;
+    uint256 totalQuantityRevoked;
+    address[] tokenHolders;
+    bool exists;
+  }
+
+  /** View information for one account's ownership of a schedule */
+  struct ScheduleDetailForAddress {
+    address tokenHolder;
+    uint256 scheduleTokenId;
+    uint256 balance;
+    uint256 claimableAmount;
+    uint256 claimedAmount;
+    uint256 quantityRevoked;
+    bool exists;
+  }
+
+  struct RestrictionDuration {
+    uint256 duration;
+    bool wasSet;
+  }
+
+  bytes32 public constant PAUSER_ROLE = keccak256("PAUSER_ROLE");
+
+  /**
+   * @notice Role conferring creation of schedules.
+   *
+   * @dev the Market contract is granted this role after deployments.
+   */
+  bytes32 public constant SCHEDULE_CREATOR_ROLE =
+    keccak256("SCHEDULE_CREATOR_ROLE");
+
+  /**
+   * @notice Role conferring sending of bpNori to this contract.
+   *
+   * @dev the Market contract is granted this role after deployments.
+   */
+  bytes32 public constant TOKEN_DEPOSITOR_ROLE =
+    keccak256("TOKEN_DEPOSITOR_ROLE");
+
+  /**
+   * @notice Role conferring revocation of restricted tokens.
+   *
+   * @dev only Nori admin address should have this role.
+   */
+  bytes32 public constant TOKEN_REVOKER_ROLE = keccak256("TOKEN_REVOKER_ROLE");
+
+  /**
+   * @notice Used to register the ERC777TokensRecipient recipient interface in the
+   * ERC-1820 registry
+   *
+   * @dev Registering that RestrictedNORI implements the ERC777TokensRecipient interface with the registry is a
+   * requirement to be able to receive ERC-777 BridgedPolygonNORI tokens. Once registered, sending BridgedPolygonNORI
+   * tokens to this contract will trigger tokensReceived as part of the lifecycle of the BridgedPolygonNORI transaction
+   */
+  bytes32 public constant ERC777_TOKENS_RECIPIENT_HASH =
+    keccak256("ERC777TokensRecipient");
+
+  mapping(uint256 => mapping(uint256 => RestrictionDuration))
+    private _methodologyAndVersionToScheduleDuration;
+
+  mapping(address => EnumerableSetUpgradeable.UintSet)
```

Comment:
honestly feel like this shouldn't exist for the same reason this type of thing doesnt exist in erc1155 and it's contributing to some poor patterns in several functions where we should be retrieving ids **off**-chain and passing them through as an array to various functions. Instead this mapping is making it convenient for several functions to that load on-chain via `storage` variables and loops

Storing it is adding a whole lot of overhead to several functions where instead we should already be able to determine the owners

Reply:
https://github.com/OpenZeppelin/openzeppelin-contracts/issues/2705

Reply:
hmm, I definitely see the point here. this information isn't required for any of our on-chain math or anything, it is just helping us understand which accounts own parts of which schedules.

so would an off-chain reconstruction of this look like getting all of the schedule ids and then, for the account in question, just querying the balance of all of them to find the ids for which the account has a balance?

Reply:
should I go ahead and remove this

Reply:
will have to change test code too since it relies on these view functions quite a bit

Reply:
> 
> so would an off-chain reconstruction of this look like getting all of the schedule ids and then, for the account in question, just querying the balance of all of them to find the ids for which the account has a balance?

Exactly.  Like the example code mentioned [here](https://github.com/OpenZeppelin/openzeppelin-contracts/issues/1102#issuecomment-799623364) to do something similar for erc721s

Beyond our own querying, I would be concerned about external smart contracts consuming this function and running into unexpected limits

Reply:
fixed